File: C++/reverseBetween.cpp
#include<iostream>

struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution{
    public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if (m == n) {
            return head;
        }

        ListNode* pre = head;
        ListNode* current = nullptr;
        ListNode* next = nullptr;

        for (int i = 0; i < m - 1; i++) {
            pre = pre->next;
        }
        current = pre->next;

        for (int i = m; i < n; i++) {
            next = current->next;
            current->next = pre;
            pre = current;
            current = next;
        }

        if (m == 1) {
            head->next = current;
            return pre;
        } else {
            ListNode* temp = head;
            for (int i = 0; i < m - 2; i++) {
                temp = temp->next;
            }
            temp->next->next = current;
            temp->next = pre;
            return head;
        }
    }
};

int main(){
    ListNode* head = new ListNode(1);
    head->next=new ListNode(2);
    head->next->next=new ListNode(3);
    head->next->next->next=new ListNode(4);
    head->next->next->next->next =new ListNode(5);
    int m=2,n=4;
    Solution solution;

    std::cout<<"Original list:";
    for(ListNode* node=head;node!=nullptr;node=node->next){
        std::cout<<node->val<<" ";
    }
    std::cout<<std::endl;

    ListNode* reversedHead = solution.reverseBetween(head,2,4);
    std::cout<<"Reversed part of list:";
    for(ListNode* node =reversedHead;node !=nullptr;node=node->next){
        std::cout<<node->val<<" ";
    }
    std::cout<<std::endl;

    while(head!=nullptr){
        ListNode* next =head->next;
        delete head;
        head=next;
    }
    
    return 0;
}

File: python/reverseBetween.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        if m == n:
            return head

        pre = head
        current = None
        next_node = None

        for i in range(0, m - 1):
            pre = pre.next
        current = pre.next

        for i in range(m, n):
            next_node = current.next
            current.next = pre
            pre = current
            current = next_node

        if m == 1:
            head.next = current
            return pre
        else:
            p = head
            for i in range(0, m - 2):
                p = p.next
            p.next.next = current
            p.next = pre
            return head

def print_list(node):
    while node:
        print(node.val, end=" -> ")
        node = node.next
    print("None")

head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(3)
head.next.next.next = ListNode(4)
head.next.next.next.next = ListNode(5)

print("Original List:")
print_list(head)

solution = Solution()
reversed_head = solution.reverseBetween(head, 2, 4)

print("List after reversing between nodes 2 to 4:")
print_list(reversed_head)